src/wasm/wasm-debug.cpp

```diff
@@ -99,26 +103,318 @@ void dumpDWARF(const Module& wasm) {
+    }
+    if (basicBlock != old.basicBlock) {
+      Fatal() << "bb";
+    }
+    if (prologueEnd != old.prologueEnd) {
+      assert(prologueEnd);
+      newOpcodes.push_back(makeItem(llvm::dwarf::DW_LNS_set_prologue_end));
+    }
+    if (epilogueBegin != old.epilogueBegin) {
+      Fatal() << "eb";
+    }
+    if (useSpecial) {
+      // Emit a special, which ends a sequence automatically.
+      // TODO
+    } else {
+      // End the sequence manually.
+      // len = 1 (subopcode)
+      newOpcodes.push_back(makeItem(llvm::dwarf::DW_LNE_end_sequence, 1));
+      // Reset the state.
+      *this = LineState(table);
+    }
+  }
+
+private:
+  llvm::DWARFYAML::LineTableOpcode makeItem(llvm::dwarf::LineNumberOps opcode) {
+    llvm::DWARFYAML::LineTableOpcode item = {};
+    item.Opcode = opcode;
+    return item;
+  }
+
+  llvm::DWARFYAML::LineTableOpcode
+  makeItem(llvm::dwarf::LineNumberExtendedOps opcode, uint64_t len) {
+    auto item = makeItem(llvm::dwarf::LineNumberOps(0));
+    // All the length after the len field itself, including the subopcode
+    // (1 byte).
+    item.ExtLen = len;
+    item.SubOpcode = opcode;
+    return item;
+  }
+};
+
+// Represents a mapping of addresses to expressions.
+struct AddrExprMap {
+  std::unordered_map<uint32_t, Expression*> map;
+
+  // Construct the map from the binaryLocations loaded from the wasm.
+  AddrExprMap(const Module& wasm) {
+    for (auto& func : wasm.functions) {
+      for (auto pair : func->binaryLocations) {
+        assert(map.count(pair.second) == 0);
+        map[pair.second] = pair.first;
+      }
+    }
+  }
+
+  // Construct the map from new binaryLocations just written
+  AddrExprMap(const BinaryLocationsMap& newLocations) {
+    for (auto pair : newLocations) {
+      assert(map.count(pair.second) == 0);
+      map[pair.second] = pair.first;
+    }
+  }
+
+  Expression* get(uint32_t addr) {
+    auto iter = map.find(addr);
+    if (iter != map.end()) {
+      return iter->second;
+    }
+    return nullptr;
+  }
+
+  void dump() {
+    std::cout << "  (size: " << map.size() << ")\n";
+    for (auto pair : map) {
+      std::cout << "  " << pair.first << " => " << pair.second << '\n';
+    }
+  }
+};
+
+static void updateDebugLines(const Module& wasm,
+                             llvm::DWARFYAML::Data& data,
+                             const BinaryLocationsMap& newLocations) {
+  // TODO: for memory efficiency, we may want to do this in a streaming manner,
+  //       binary to binary, without YAML IR.
+
+  // TODO: apparently DWARF offsets may be into the middle of instructions...
+  //       we may need to track their spans too
+  // https://github.com/WebAssembly/debugging/issues/9#issuecomment-567720872
+  AddrExprMap oldAddrMap(wasm);
+  // std::cout << "old\n";
+  // oldAddrMap.dump();
+  AddrExprMap newAddrMap(newLocations);
+  // std::cout << "new\n";
+  // newAddrMap.dump();
+
+  for (auto& table : data.DebugLines) {
```

Comment:
does this mean there can be more than one `.debug_line` section in a binary?
I guess this structure could also be useful if we have split debug info...

Reply:
I haven't read enough in the DWARF spec to know... but all the LLVM code supports multiple tables. Each table can have different defaults and settings, maybe that's the main use?